compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -148,6 +149,47 @@ fn check_union_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) -> b
     true
 }
 
+/// Check that the unsafe fields do not need dropping.
+fn check_unsafe_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) {
+    let item_type = tcx.type_of(item_def_id).instantiate_identity();
+    let ty::Adt(def, args) = item_type.kind() else {
+        span_bug!(span, "structs/enums must be ty::Adt, but got {:?}", item_type.kind());
+    };
+    let typing_env = ty::TypingEnv::non_body_analysis(tcx, item_def_id);
+    for variant in def.variants() {
+        for field in &variant.fields {
+            if field.safety != Safety::Unsafe {
+                continue;
+            }
+            let Ok(field_ty) = tcx.try_normalize_erasing_regions(typing_env, field.ty(tcx, args))
+            else {
+                tcx.dcx().span_delayed_bug(span, "could not normalize field type");
+                continue;
+            };
+
+            if !allowed_union_or_unsafe_field(field_ty, tcx, typing_env) {
```

Comment:
This is an existing function I moved out of a nested scope - I have a mild preference for not changing it too much here :-)

Reply:
It has four callsites all of which you've touched already because you renamed the function, so you're not really avoiding anything by fixing this on the way lol

Reply:
fair :-)